tern/analyze/docker/container.py

```diff
@@ -192,3 +198,139 @@ def close_client():
         # it should either already be closed, no socker is in use,
         # or docker is not setup -- either way, the socket is closed
         pass
+
+
+def save_image(image_tag_string):
+    '''Download and save a docker image with given name and tag '''
+    rootfs.create_working_dir()
+    temp_path = os.path.join(rootfs.mount_dir, temp_folder)
+    image_name, image_tag = image_tag_string.split(":")
+    manifest = pull_image_manifest(image_name, image_tag)
+    manifest_path = os.path.join(temp_path, "manifest.json")
+    # saving image manifest
+    json.dump(manifest, open(manifest_path, "w"))
+
+    digest = manifest.get("config").get("digest")
+    config = pull_image_config(image_name, image_tag, digest)
+    config_name = digest.split(":")[1]
+    config_path = os.path.join(temp_path, config_name)
+    # saving image config
+    json.dump(config, open(config_path, "w"))
+
+    # saving image layers
+    layers = manifest.get("layers")
+    for layer, content in pull_image_layers(image_name, image_tag, layers):
+        layer_path = os.path.join(temp_path, layer)
+        with open(layer_path, "wb") as layer_content:
+            layer_content.write(content)
+
+    return temp_path
+
+
+def generate_manifest_url(image_name, image_tag):
+    '''
+    Generate a docker image manifest url end point with
+    given image name and tag
+    '''
+    url = "{0}/{1}/manifests/{2}"
+    return url.format(docker_endpoint, image_name, image_tag)
+
+
+def generage_image_blobs_url(image_name, digest):
+    '''
+    Generate a docker image blobs url end point with
+    given image name and tag
+    '''
+    url = "{0}/{1}/blobs/{2}"
+    return url.format(docker_endpoint, image_name, digest)
+
+
+def is_token_expired():
```

Comment:
I think this bearer authentication logic can be placed in `tern/utils/general.py` as it can be used for endpoints other than Docker.

Reply:
`is_token_expired(auth_token)` so you don't need to use a global `auth_token` variable.

Reply:
@nishakm here different endpoint means container services apart from docker you meant? I think getting auth token could be different for other endpoints in that case.

Reply:
Definitely, but you would still have implemented the Bearer Authentication method. Incidentally, `requests` has a way of adding new auth methods: https://requests.readthedocs.io/en/master/user/authentication/#new-forms-of-authentication